src/TraceEvent/EventPipe/EventPipeEventSource.cs

```diff
@@ -58,80 +90,89 @@ protected override void Dispose(bool disposing)
-                    Dispatch(event_);
-                    sessionEndTimeQPC = eventRecord->EventHeader.TimeStamp;
+                    TraceEventNativeMethods.EVENT_RECORD* eventRecord = ReadEvent(deserializerReader);
+                    if (eventRecord != null)
+                    {
+                        // in the code below we set sessionEndTimeQPC to be the timestamp of the last event.  
+                        // Thus the new timestamp should be later, and not more than 1 day later.  
+                        Debug.Assert(sessionEndTimeQPC <= eventRecord->EventHeader.TimeStamp);
+                        Debug.Assert(sessionEndTimeQPC == 0 || eventRecord->EventHeader.TimeStamp - sessionEndTimeQPC < _QPCFreq * 24 * 3600);
+
+                        var traceEvent = Lookup(eventRecord);
+                        Dispatch(traceEvent);
+                        sessionEndTimeQPC = eventRecord->EventHeader.TimeStamp;
+                    }
                 }
             }
-
+#endif
             return true;
         }
 
-        internal override string ProcessName(int processID, long timeQPC)
-        {
-            return _processName;
-        }
+        internal override string ProcessName(int processID, long timeQPC) => _processName;
 
-        private TraceEventNativeMethods.EVENT_RECORD* ReadEvent(PinnedStreamReader reader)
+        internal TraceEventNativeMethods.EVENT_RECORD* ReadEvent(PinnedStreamReader reader)
         {
-            // Guess that the event is < 1000 bytes or whatever is left in the stream.  
-            int eventSizeGuess = Math.Min(1000, _endOfEventStream.Sub(reader.Current));
+            nint eventSizeGuess =
+                _endOfEventStream != 0 // V3 has no forward reference to the end
+                    ? Math.Min(1000, _endOfEventStream.Sub(reader.Current)) // Guess that the event is < 1000 bytes or whatever is left in the stream.  
+                    : sizeof(EventPipeEventHeader); // the minimum size (we can't use 1000 because it might be more than the file's length)
+
             EventPipeEventHeader* eventData = (EventPipeEventHeader*)reader.GetPointer(eventSizeGuess);
+            
             // Basic sanity checks.  Are the timestamps and sizes sane.  
             Debug.Assert(sessionEndTimeQPC <= eventData->TimeStamp);
             Debug.Assert(sessionEndTimeQPC == 0 || eventData->TimeStamp - sessionEndTimeQPC < _QPCFreq * 24 * 3600);
             Debug.Assert(0 <= eventData->PayloadSize && eventData->PayloadSize <= eventData->TotalEventSize);
-            Debug.Assert(eventData->MetaDataId <= reader.Current);       // IDs are the location in the of of the data, so it comes before
             Debug.Assert(0 < eventData->TotalEventSize && eventData->TotalEventSize < 0x20000);  // TODO really should be 64K but BulkSurvivingObjectRanges needs fixing.
 
+            nint eventDataEndIncludingAlignmentPadding = (nint)reader.Current + eventData->TotalEventSize;
```

Comment:
This should be 

StreamLabel eventDataEndIncludingAlignmentPadding = reader.Current.Add(eventData->TotalEventSize);